runners/samza/src/main/java/org/apache/beam/runners/samza/metrics/BeamTransformMetricRegistry.java

```diff
@@ -0,0 +1,129 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.runners.samza.metrics;
+
+import java.io.Serializable;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
+import org.apache.samza.context.Context;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * BeamTransformMetricRegistry is a registry that maintains the metrics for each transform. It
+ * maintains the average arrival time for each PCollection for a primitive transform.
+ *
+ * <p>For a non-data shuffling primitive transform, the average arrival time is calculated per
+ * watermark, per PCollection {@link org.apache.beam.sdk.values.PValue} and updated in
+ * avgArrivalTimeMap
+ */
+public class BeamTransformMetricRegistry implements Serializable {
+  private static final Logger LOG = LoggerFactory.getLogger(BeamTransformMetricRegistry.class);
+
+  // TransformName -> PValue for pCollection -> Map<WatermarkId, AvgArrivalTime>
+  private ConcurrentHashMap<String, ConcurrentHashMap<String, ConcurrentHashMap<Long, Long>>>
```

Comment:
this is done because we need to keep the arrival time per pValue at entry and exit of transform, this is because if a dag splits out then an element traversers left sub-branch and then traverses the right child branch 


![image](https://user-images.githubusercontent.com/16806974/235499463-777aa2b1-4200-471f-bb5c-e2f59a6c1286.png)

This was brought up during the review, can explain offline

Reply:
In essense, we are keeping track of each PCollection's timestamp, regardless of using per transform or not. In the above graph, we should use the latency due to traversing 4 first. Otherwise we are not reporting correctly the latency. In another world we might do some optimization to execute 4 and 5 in parallel, then we won't be able see the latency improvement. I think this should be a simple equation,  basically `timestamp(input PC) - timestamp(output PC)` for each PTransform.

Reply:
Synced offline explained why it is important to attach Metric before & after every transform